clvm/ConvertingCLVMObject.py

```diff
@@ -0,0 +1,108 @@
+from blspy import G1Element
+from .casts import int_to_bytes
+from .CLVMObject import CLVMObject
+
+
+def looks_like_clvm_object(o: Any) -> bool:
+    d = dir(o)
+    return "atom" in d and "pair" in d
+
+
+CastableType = Union[
+    "CLVMObject",
+    bytes,
+    int,
+    None,
+    G1Element,
+    Tuple[Any, Any],
+]
+
+
+class ConvertingCLVMObject:
+    """
+    This class implements the CLVM Object protocol, but supports deferring
+    conversion of python types until they are acceessed. For example, it can hold a
+    G1Element, which is converted to bytes the first time 'atom' is accessed. It can
+    hold a list, which is converted into a cons-list one element at a time, when
+    iterated.
+    """
+
+    # if this is not bytes, it will be lazily converted to bytes when accessed
+    atom_: Optional[Any]
+
+    # this is always a 2-tuple, but the elements may not have been converted to
+    # a CLVM object, they need to be when accessed. If an element is already a
+    # valid clvm object, it should be left alone. Otherwise it should be wrapped
+    # in CLVMObject to defer conversion when accessed
+    pair_: Optional[Tuple[Any, Any]]
+    __slots__ = ["atom_", "pair_"]
+
+    def __new__(class_, v: CastableType):
+        if isinstance(v, ConvertingCLVMObject):
+            return v
+
+        # TODO: ideally we wouldn't mix SExp into the tree like this
+        if type(v).__name__ in ["SExp", "Program"]:
+            return v.obj_
+
+        self = super(ConvertingCLVMObject, class_).__new__(class_)
+        if isinstance(v, tuple):
+            if len(v) != 2:
+                raise ValueError("tuples must be of size 2, cannot create ConvertingCLVMObject from: %s" % str(v))
+            self.pair_ = v
+            self.atom_ = None
+        elif isinstance(v, list):
+            if len(v) == 0:
+                self.atom_ = b""
+                self.pair_ = None
+            else:
+                # lazily convert the list into the tree structure
+                self.pair_ = (v[0], v[1:])
+                self.atom_ = None
+        elif v is None:
+            self.atom_ = b""
+            self.pair_ = None
+        else:
+            self.atom_ = v
+            self.pair_ = None
+        return self
+
+    @property
+    def atom(self):
+        if self.atom_ is not None and not isinstance(self.atom_, bytes):
+            # deferred conversion
+            self.atom_ = convert_atom_to_bytes(self.atom_)
+        return self.atom_
+
+    @property
+    def pair(self):
+        if self.pair_ is not None:
+            valid_first = isinstance(self.pair_[0], ConvertingCLVMObject) or looks_like_clvm_object(self.pair_[0])
+            valid_second = isinstance(self.pair_[1], ConvertingCLVMObject) or looks_like_clvm_object(self.pair_[1])
+            if not valid_first or not valid_second:
+                # deferred conversion
+                self.pair_ = (self.pair_[0] if valid_first else ConvertingCLVMObject(self.pair_[0]),
```

Comment:
maybe there's a way to assign `self.pair` in here, and `self.atom` in the other property, that would make any future attempts to access these fields just go straight to the members, rather than these functions.

Reply:
```python
class foo:
    @property
    def val(self):
        self.val = 100
        return 200

f1 = foo()
print(f1.val)
print(f1.val)
print(f1.val)
```

This prints
```
200
100
100
100
```

so there's your hack.

Reply:
I can't get this to work. I've tried `self.pair = `, `setattr(self, "pair", ...)`, `super().__setattr__("pair", ...)`. They all fail with:
```
AttributeError: can't set attribute
```

Presumably because `pair` is still treated as a property that doesn't have a setter, just a getter.

The one thing that didn't fail was `self.__dict__["pair"] =`, but that did not override the `@property`, so it didn't do any good.

I imagine that perhaps this is related to `__slots__ = ["atom", "pair_"]`. That was one of the first things I removed, to enable "objects are arbitrary dicts" behavior. I also suspected it had something to do with the existence if `__new__()`. I tried to replace that with an `__init__()`, but that also did not make a difference.

Reply:
It turns out I was (unwittingly) running it under `python2`.